lib/src/io.dart

```diff
@@ -788,102 +791,87 @@ Future<HttpServer> bindServer(String host, int port) async {
-  for (final file in ['/bin/tar', '/usr/bin/tar']) {
-    if (fileExists(file)) {
-      return file;
-    }
-  }
-  log.warning(
-      'Could not find a system `tar` installed in /bin/tar or /usr/bin/tar, '
-      'attempting to use tar from PATH');
-  return 'tar';
-}
-
 /// Extracts a `.tar.gz` file from [stream] to [destination].
 Future extractTarGz(Stream<List<int>> stream, String destination) async {
   log.fine('Extracting .tar.gz stream to $destination.');
-  final decompressed = stream.transform(GZipCodec().decoder);
 
-  // We used to stream directly to `tar`,  but that was fragile in certain
-  // settings.
-  final processResult = await withTempDir((tempDir) async {
-    final tarFile = path.join(tempDir, 'archive.tar');
-    try {
-      await _createFileFromStream(decompressed, tarFile);
-    } catch (e) {
-      // We don't know the error type here: https://dartbug.com/41270
-      throw FileSystemException('Could not decompress gz stream $e');
+  destination = path.absolute(destination);
+  final reader = TarReader(stream.transform(gzip.decoder));
+  while (await reader.moveNext()) {
+    final entry = reader.current;
+
+    final filePath = path.joinAll([
+      destination,
+      // Tar file names always use forward slashes
+      ...path.posix.split(entry.name),
+    ]);
+
+    if (!path.isWithin(destination, filePath)) {
+      // The tar contains entries that would be written outside of the
+      // destination. That doesn't happen by accident, assume that the tar file
+      // is malicious.
+      await reader.cancel();
+      throw FormatException('Invalid tar entry: ${entry.name}');
+    }
+
+    final parentDirectory = path.dirname(filePath);
+
+    switch (entry.type) {
+      case TypeFlag.dir:
+        ensureDir(filePath);
+        break;
+      case TypeFlag.reg:
+      case TypeFlag.regA:
+        // Regular file
+        deleteIfLink(filePath);
+        ensureDir(parentDirectory);
+
+        await _createFileFromStream(entry.contents, filePath);
+
+        if (Platform.isLinux || Platform.isMacOS) {
+          // Apply executable bits from tar header, but don't change r/w bits
+          // from the default
+          const defaultMode = 420; // 644 in base 8
+          const executableMask = 0x49; // 001 001 001
+          final mode = defaultMode | (entry.header.mode & executableMask);
+
+          if (mode != defaultMode) {
+            _chmod(mode, filePath);
+          }
+        }
+        break;
+      case TypeFlag.symlink:
+        // Link to another file in this tar, relative from this entry.
+        final resolvedTarget = path.joinAll(
+            [parentDirectory, ...path.posix.split(entry.header.linkName)]);
+        if (!path.isWithin(destination, resolvedTarget)) {
+          // Don't allow links to files outside of this tar.
+          break;
+        }
+
+        ensureDir(parentDirectory);
+        createSymlink(path.relative(resolvedTarget, from: destination), filePath);
```

Comment:
Sorry - my suggestion was wrong! It should be (like you wrote below):
```suggestion
        createSymlink(path.relative(resolvedTarget, from: parentDirectory), filePath);
```

Maybe it is worth extracting the link-making code that is shared between the two types of link.
Or you could even make an empty case that falls through:

```dart
    case TypeFlag.link:
        // We generate hardlinks as symlinks too, but their path needs to be
        // resolved against the root of the tar file.
    case TypeFlag.symlink:
      final resolvedTarget = path.join(parentDirectory, entry.header.linkName);
      if (!path.isWithin(destination, resolvedTarget)) {
          break; // Link points outside of the tar file.
      }
      ensureDir(parentDirectory);
      createSymlink(resolvedTarget, filePath, relative: true);
      break;
```

Reply:
> Or you could even make an empty case that falls through:

I'm not sure that this would work? For `TypeFlag.symlink`, `header.linkName` is relative to the entry. For `TypeFlag.link`, it's relative to the root of the tar file.

That's at least what GNU-tar appears to do, I couldn't find this documented anywhere:

```
mkdir -p tar/source
touch tar/source/source.txt
mkdir tar/links
cd tar/links
ln ../source/source.txt hardlink.txt
ln -s ../source/source.txt softlink.txt
cd ../..
tar --create --file out.tar tar/
```

GNU tar generates `links/softlink.txt` as a softlink with `entry.header.linkName` pointing to `../source/source.txt`. It then generates the hardlink as a regular file and `tar/source/source.txt` as a hardlink with a `linkName` of `tar/links/hardlink.txt`.

I couldn't find a way to merge the two cases without making the code more complicated.

Reply:
Ah I was unaware!

Ok then the cases should indeed not be merged!

We could add a comment here explaining the difference.

We should also have a test that the resulting targets of both kinds of symlinks (and perhaps both for relative and absolute) are correct to avoid someone later "cleaning" this up.

Reply:
> We could add a comment here explaining the difference.

Well, we have this:

```dart
        // We generate hardlinks as symlinks too, but their path needs to be
        // resolved against the root of the tar file.
```

Clearly that didn't clear things up enough, I'll expand the comment.

> We should also have a test that the resulting targets of both kinds of symlinks

I've added a test for that here: https://github.com/dart-lang/pub/blob/4928b5b456205150ffb927eb61bb8c9643c05494/test/io_test.dart#L310-L348

Both links are in subdirectories themselves, so the test would file if either case was merged into the other. If you want a more specific test case let me know.

Reply:
No that is fine - I somehow overlooked those!